src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs

```diff
@@ -7314,32 +7343,48 @@ class C
             delegate void X(ref readonly int p);
             delegate void Y({{modifier}} int p);
             """;
-        CompileAndVerify(source, expectedOutput: "XYYX").VerifyDiagnostics(
-            // (8,5): warning CS9198: Reference kind modifier of parameter 'in int p' doesn't match the corresponding parameter 'ref readonly int p' in target.
-            // x = (X)C.Y;
-            Diagnostic(ErrorCode.WRN_TargetDifferentRefness, "(X)C.Y").WithArguments($"{modifier} int p", "ref readonly int p").WithLocation(8, 5),
-            // (9,5): warning CS9198: Reference kind modifier of parameter 'ref readonly int p' doesn't match the corresponding parameter 'in int p' in target.
-            // y = (Y)C.X;
-            Diagnostic(ErrorCode.WRN_TargetDifferentRefness, "(Y)C.X").WithArguments("ref readonly int p", $"{modifier} int p").WithLocation(9, 5));
-    }
 
-    [Theory, CombinatorialData]
-    public void Conversion_DelegateConstruction([CombinatorialValues("in", "ref")] string modifier)
-    {
-        var source = $$"""
+        var source2 = """
             X x = C.X;
-            Y y = C.Y;
-
             var i = 1;
             x(in i);
-            y({{modifier}} i);
+            x = (X)C.Y;
+            x(in i);
+            """;
 
-            x = new X(C.Y);
-            y = new Y(C.X);
+        var source3 = $"""
+            Y y = C.Y;
+            var i = 1;
+            y({modifier} i);
+            y = (Y)C.X;
+            y({modifier} i);
+            """;
 
-            x(in i);
-            y({{modifier}} i);
+        CompileAndVerify(new[] { source1, source3 }, expectedOutput: "YX").VerifyDiagnostics(
+            // 1.cs(4,5): warning CS9198: Reference kind modifier of parameter 'ref readonly int p' doesn't match the corresponding parameter 'in int p' in target.
+            // y = (Y)C.X;
+            Diagnostic(ErrorCode.WRN_TargetDifferentRefness, "(Y)C.X").WithArguments("ref readonly int p", $"{modifier} int p").WithLocation(4, 5));
 
+        if (modifier == "ref")
+        {
+            CreateCompilation(new[] { source1, source2 }).VerifyDiagnostics(
+                // 1.cs(4,5): error CS0123: No overload for 'Y' matches delegate 'X'
+                // x = (X)C.Y;
+                Diagnostic(ErrorCode.ERR_MethDelegateMismatch, "(X)C.Y").WithArguments("Y", "X").WithLocation(4, 5));
+        }
+        else
+        {
+            CompileAndVerify(new[] { source1, source2 }, expectedOutput: "XY").VerifyDiagnostics(
+                // 1.cs(4,5): warning CS9198: Reference kind modifier of parameter 'in int p' doesn't match the corresponding parameter 'ref readonly int p' in target.
+                // x = (X)C.Y;
+                Diagnostic(ErrorCode.WRN_TargetDifferentRefness, "(X)C.Y").WithArguments("in int p", "ref readonly int p").WithLocation(4, 5));
+        }
```

Comment:
Can we move this section between the definition of `source2` and `source3`? That puts the code closer to the diagnostics and makes it a bit easier to read. 

Similar feedback for later tests.